grails-app/domain/org/pih/warehouse/inventory/CycleCount.groovy

```diff
@@ -0,0 +1,64 @@
+package org.pih.warehouse.inventory
+
+import org.pih.warehouse.auth.AuthService
+import org.pih.warehouse.core.Location
+import org.pih.warehouse.core.User
+
+class CycleCount {
+
+    String id
+
+    Location facility
+
+    Date dateLastRefreshed
+
+    Date dateCreated
+
+    Date lastUpdated
+
+    User createdBy
+
+    User updatedBy
+
+    def beforeInsert() {
+        createdBy = AuthService.currentUser
+        updatedBy = AuthService.currentUser
+    }
+
+    def beforeUpdate() {
+        updatedBy = AuthService.currentUser
+    }
+
+    static constraints = {
+        id generator: "uuid"
+        createdBy(nullable: true)
+        updatedBy(nullable: true)
+    }
+
+    CycleCountStatus getStatus() {
+        List<CycleCountItem> cycleCountItems = CycleCountItem.findAllByCycleCount(this)
+        if (cycleCountItems.every { it.status == CycleCountItemStatus.READY_TO_COUNT }) {
+            return CycleCountStatus.REQUESTED
+        }
+        if (cycleCountItems.every { it.status == CycleCountItemStatus.REVIEWED }) {
+            return CycleCountStatus.REVIEWED
+        }
+        if (cycleCountItems.every { it.status == CycleCountItemStatus.TO_REVIEW }) {
+            return CycleCountStatus.TO_REVIEW
+        }
+        if (cycleCountItems.any { it.status == CycleCountItemStatus.INVESTIGATING }) {
+            return CycleCountStatus.INVESTIGATING
+        }
+        if (cycleCountItems.any { it.status == CycleCountItemStatus.COUNTING }
+                && !cycleCountItems.any { it.status in [CycleCountItemStatus.INVESTIGATING, CycleCountItemStatus.TO_REVIEW] }) {
+            return CycleCountStatus.COUNTING
+        }
+        if (cycleCountItems.every { it.status in [CycleCountItemStatus.COUNTED, CycleCountItemStatus.TO_REVIEW] }) {
+            return CycleCountStatus.COUNTED
+        }
+        if (cycleCountItems.every { it.status in [CycleCountItemStatus.REVIEWED, CycleCountItemStatus.APPROVED, CycleCountItemStatus.REJECTED] }) {
+            return CycleCountStatus.COMPLETED
+        }
+        return null
```

Comment:
I followed this:
![Screenshot from 2025-01-28 10-08-02](https://github.com/user-attachments/assets/12731e1c-4212-4cf6-8f85-0a16b3adec7d)